### PR TITLE
Variant shredding

### PR DESCRIPTION
**Pull request by cashmand**

### CODE REVIEW

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+        ArrayList<VariantBuilder.FieldEntry> fields =
+            ((VariantObjectConverter) converters[variantSchema.typedIdx]).getFields();
+
+        VariantUtil.handleObject(value, 0, (info) -> {
+          for (int i = 0; i < info.numElements; ++i) {
+            int id = VariantUtil.readUnsigned(value, info.idStart + info.idSize * i, info.idSize);
+            int offset = VariantUtil.readUnsigned(
+                value, info.offsetStart + info.offsetSize * i, info.offsetSize);
+            int elementPos = info.dataStart + offset;
+            String key = VariantUtil.getMetadataKey(this.builder.getMetadata().getBytes(), id);
```
Comment (cashmand) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java]: I think this could copy the metadata binary for every key lookup. Probably want a version of getMetadataKey that takes a ByteBuffer or Binary.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+    }
+    if (variantValue != null) {
+      if (startWritePos == builder.getWritePos()) {
+        // Nothing else was added. We can directly append this value.
+        builder.shallowAppendVariant(
+            variantValue.toByteBuffer().array(),
+            variantValue.toByteBuffer().position());
+      } else {
+        // Both value and typed_value were non-null. This is only valid for an object.
+        byte[] value = variantValue.getBytes();
```
Comment (cashmand) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:220]: Probably copies `value` in memory unnecessarily.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+  @Override
+  public void init(VariantBuilderHolder builderHolder) {
+    builder = builderHolder;
+  }
+
+  @Override
+  public void addBinary(Binary value) {
+    // TODO: create a separate converter class for each type, or find some other way to avoid these checks?
+    if (scalarType instanceof VariantSchema.StringType) {
+      builder.builder.appendString(value.toStringUsingUTF8());
```
Comment (cashmand) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java]: Should we consider making raw versions of some of these builder methods? I'm probably converting this binary to a string, only to have VariantBuilder convert it back to bytes.

File: parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java [hunk 4]
```diff
@@ -147,6 +147,12 @@ protected LogicalTypeAnnotation fromString(List<String> params) {
         return float16Type();
       }
     },
+    VARIANT {
```
Comment (rdblue) [inline parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java]: This should be working because it was added in https://github.com/apache/parquet-java/pull/3072

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroConverters.java [hunk 5]
```diff
@@ -363,4 +365,27 @@ public String convert(Binary binary) {
       return stringifier.stringify(binary);
     }
   }
+
+  static final class FieldVariantConverter<T> extends VariantColumnConverter {
```
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroConverters.java:369]: Don't we want to produce a `Variant` from https://github.com/apache/parquet-java/pull/3072 rather than a record with a record determined by the schema?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java [hunk 6]
```diff
@@ -383,7 +386,15 @@ private static Converter newConverter(
         }
         return newStringConverter(schema, model, parent);
       case RECORD:
-        return new AvroRecordConverter(parent, type.asGroupType(), schema, model);
+        // TODO: remove first condition. Just using it for testing until we have a real logical type annotation.
```
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java]: The type annotation is in, so you can use it now.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantSchema.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+/**
+ * Defines a valid shredding schema, as described in
+ * https://github.com/apache/parquet-format/blob/master/VariantShredding.md.
+ * A shredding schema contains a value and optional typed_value field.
+ * If a typed_value is an array or struct, it recursively contain its own shredding schema for
+ * elements and fields, respectively.
+ * The schema also contains a metadata field at the top level, but not in recursively shredded
+ * fields.
+ */
+public class VariantSchema {
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantSchema.java]: I don't think it's a good idea to have a Parquet schema, a Variant type system, and a shredded variant schema. Can't this use Parquet types directly like Iceberg does?

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantShreddingSchemaConverter.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+import org.apache.parquet.schema.GroupType;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
+import org.apache.parquet.schema.Type;
+
+/**
+ * <p>
+ * Converts a VariantSchema into a Parquet schema, or vice versa.
+ * </p>
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantShreddingSchemaConverter.java]: Javadoc style typically omits closing tags where they can be inferred.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+import org.apache.parquet.variant.Variant;
+import org.apache.parquet.variant.VariantBuilder;
+import org.apache.parquet.variant.VariantUtil;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestVariant extends DirectWriterTest {
+
+  @Test
+  public void testUnshredded() throws Exception {
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java:197]: I think it's fine to have tests that directly exercise the backing column API like this... but it's a huge pain to build, review, and maintain them. I would recommend instead using a system like we have in Iceberg, where we use Avro records to write all of the data, then interpret the file using a variant type and then assert that the variant value matches the expected. See test cases here: https://github.com/apache/iceberg/blob/main/parquet/src/test/java/org/apache/iceberg/parquet/TestVariantReaders.java#L161-L178

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroConverters.java (hunk 5 above)
Comment (cashmand) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroConverters.java:369]: I'm not sure I understand the comment. I'm not that familiar with Avro, but doesn't the type we produce need to be an Avro type? What would it mean to produce a `Variant`?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 10]
```diff
@@ -0,0 +1,1063 @@
...
+
+  // We need to store two copies of the schema: one without the Variant type annotation that is used to construct the
+  // Avro schema for writing, and one with type annotation that is used in the actual written parquet schema, and when
+  // reading.
+  private class TestSchema {
+    MessageType parquetSchema;
+    MessageType unannotatedParquetSchema;
+    GroupType variantType;
+    GroupType unannotatedVariantType;
+    TestSchema(Type shreddedType) {
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java:596]: Needs a newline between the constructor and the instance fields.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 11]
```diff
@@ -0,0 +1,1063 @@
...
+      }
+      Assert.assertNull(
+          "Should only contain " + expectedRecords.length + " record" + (expectedRecords.length == 1 ? "" : "s"),
+          reader.read());
+  }
+
+  // We need to store two copies of the schema: one without the Variant type annotation that is used to construct the
+  // Avro schema for writing, and one with type annotation that is used in the actual written parquet schema, and when
+  // reading.
+  private class TestSchema {
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: This should be static.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 12]
```diff
@@ -0,0 +1,1063 @@
...
+    }
+
+    TestSchema() {
+      variantType = variant("var", 2);
+      unannotatedVariantType = unannotatedVariant("var", 2);
+      parquetSchema = parquetSchema(variantType);
+      unannotatedParquetSchema = parquetSchema(unannotatedVariantType);
+    }
+
+    TestSchema(GroupType variantType, GroupType unannotatedVariantType, MessageType parquetSchema, MessageType unannotatedParquetSchema) {
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Aren't the Parquet schemas just the variant types wrapped using `parquetSchema`? Why pass them in?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 13]
```diff
@@ -0,0 +1,1063 @@
...
+  }
+
+  // The following tests are based on Iceberg's TestVariantReaders suite.
+  @Test
+  public void testUnshreddedVariants() throws Exception {
+    for (Variant v : PRIMITIVES) {
+      TestSchema schema = new TestSchema();
+
+      GenericRecord variant = recordFromMap(schema.unannotatedVariantType,
+          ImmutableMap.of("metadata", serialize(v.getMetadata()), "value", serialize(v.getValue())));
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Aren't the results of `getMetadata` and `getValue` buffers? Do you need the `serialize` wrapper?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 14]
```diff
@@ -0,0 +1,1063 @@
...
+      GenericRecord record = recordFromMap(schema.unannotatedParquetSchema, ImmutableMap.of("id", 1, "var", variant));
+      GenericRecord actual = writeAndRead(schema, record);
+      Assert.assertEquals(actual.get("id"), 1);
+
+      GenericRecord actualVariant = (GenericRecord) actual.get("var");
+      Assert.assertEquals(ByteBuffer.wrap(v.getValue()), (ByteBuffer) actualVariant.get("value"));
+      Assert.assertEquals(ByteBuffer.wrap(v.getMetadata()), (ByteBuffer) actualVariant.get("metadata"));
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: This test and the one above look good.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 15]
```diff
@@ -0,0 +1,1063 @@
...
+      TestSchema schema = new TestSchema(shreddedType(v));
+
+      GenericRecord variant =
+          recordFromMap(
+              schema.unannotatedVariantType,
+              ImmutableMap.of(
+                  "metadata",
+                  v.getMetadata(),
+                  "typed_value",
+                  toAvroValue(v)));
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Could your variant test case code produce both the value and the variant equivalent? That would make this easier, rather than needing to validate `toAvroValue(v)`.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 16]
```diff
@@ -0,0 +1,1063 @@
...
+      GenericRecord record = recordFromMap(schema.unannotatedParquetSchema, ImmutableMap.of("id", 1, "var", variant));
+
+      GenericRecord actual = writeAndRead(schema, record);
+      Assert.assertEquals(actual.get("id"), 1);
+
+      GenericRecord actualVariant = (GenericRecord) actual.get("var");
+      Assert.assertEquals(ByteBuffer.wrap(v.getValue()), (ByteBuffer) actualVariant.get("value"));
+      Assert.assertEquals(ByteBuffer.wrap(v.getMetadata()), (ByteBuffer) actualVariant.get("metadata"));
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: These assertions work for the two test cases above because I expect Parquet not to modify the `ByteBuffer` values. However, for this case I think the assertion needs to validate the semantic value of the variant. That is, the actual value should have the expected value's type and be equal to the expected value.

That's because this is a stronger assertion than semantic value. For instance, there are two ways to encode a variant string, as a short string or as a primitive with a string value. This test doesn't need to validate that the short string is used and that the bytes match exactly.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java (hunk 12 above)
Comment (rdblue) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Nevermind, I see that this is for cases where we are exercising edge cases, like a missing `value` column.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java [hunk 17]
```diff
@@ -0,0 +1,1063 @@
...
+    GenericRecord record = recordFromMap(schema.unannotatedParquetSchema, ImmutableMap.of("id", 1, "var", variant));
+
+    assertThrows(() -> writeAndRead(schema, record),
+        UnsupportedOperationException.class,
+        "Unsupported shredded value type: optional fixed_len_byte_array(4) typed_value");
+  }
+
+  @Test
+  public void testShreddedObject() throws IOException {
+    // TODO: Build this at class level, without parseJson, need to deal with exception.
```
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Having an assertion that matches the structure would help out here as well.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java (hunk 13 above)
Comment (cashmand) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: You're right, I'll remove them in the final PR. I had been trying to minimize the edits I had to make after copying each test from iceberg.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java (hunk 16 above)
Comment (cashmand) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: I agree that this is a stronger assertion than needed. The fact that it never fails makes me think that we should probably add some test cases where it is violated - e.g. I think my code should produce a different binary depending on the order of object fields in the parquet schema; we should probably add some tests where they aren't alphabetically ordered.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java (hunk 15 above)
Comment (cashmand) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: I agree that I don't love `toAvroValue`, but I'm not sure what you're suggesting. Do you mean write out a pair for every primitive test case, like `new PrimitiveCase(12345, b -> b.appendLong(12345)`?

Comment (rdblue) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestVariant.java]: Yes, that's exactly it. That way we can easily see that the value is correct, rather than converting to a variant and back.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroIndexedRecordConverter.java [hunk 18]
```diff
@@ -168,7 +169,12 @@ private static Converter newConverter(Schema schema, Type type, GenericData mode
       case MAP:
         return new MapConverter(parent, type.asGroupType(), schema, model);
       case RECORD:
-        return new AvroIndexedRecordConverter<>(parent, type.asGroupType(), schema, model);
+        if (type.getLogicalTypeAnnotation()
+                instanceof LogicalTypeAnnotation.VariantLogicalTypeAnnotation) {
```
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroIndexedRecordConverter.java:173]: Continuation indent should be 2 indents (4 spaces)

File: parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java [hunk 19]
```diff
@@ -708,6 +708,7 @@ public Builder withStatisticsEnabled(String columnPath, boolean enabled) {
     }
 
     public Builder withStatisticsEnabled(boolean enabled) {
+      this.statistics.withDefaultValue(enabled);
```
Comment (rdblue) [inline parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java:711]: Why is this change needed?

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 20]
```diff
@@ -0,0 +1,799 @@
...
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.BOOLEAN;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.DOUBLE;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.FLOAT;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT32;
+import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.INT64;
+import static org.apache.parquet.schema.Type.Repetition.REPEATED;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.*;
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:32]: We avoid wildcard imports because imports are generally managed by an IDE and it helps to be able to grep for them or to see where symbols are coming from in github.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 21]
```diff
@@ -0,0 +1,799 @@
...
+}
+
+interface VariantConverter {
+  void init(VariantBuilderHolder builderHolder);
+}
+
+/**
+ * Converter for a shredded Variant containing a value and/or typed_value field: either a top-level
+ * Variant column, or a nested array element or object field.
+ * The top-level converter is handled by a subclass (VariantColumnConverter)  that also reads
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:86]: Looks like odd spacing and line wrapping.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 22]
```diff
@@ -0,0 +1,799 @@
...
+ * metadata.
+ *
+ * All converters for a Variant column shared the same VariantBuilder, and append their results to
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:89]: Typo: "shared" -> "share"

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 23]
```diff
@@ -0,0 +1,799 @@
...
+    // rows with identical metadata should be the most common case.
+    if (this.metadata != metadata) {
+      metadataMap = VariantUtil.getMetadataMap(metadata.getBytes());
+    }
+    this.metadata = metadata;
+    builder.setFixedMetadata(metadataMap);
+  }
+}
+
+interface VariantConverter {
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:79]: In Java, the convention is to add top-level classes in their own files that have the same name as the class or interface. If you only need a class or interface locally, you would normally nest it in the class that the file is named for, `VariantColumnConverter`.

It's not uncommon to have files with only a single small type, like this interface.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 24]
```diff
@@ -0,0 +1,799 @@
...
+ * appended by the `typed_value` converter, but because residual values are stored in `value`, this
+ * converter is responsible for finalizing the object.
+ */
+class VariantElementConverter extends GroupConverter implements VariantConverter {
+
+  // startWritePos has two uses:
+  // 1) If typed_value is an object, we gather fields from value and typed_value and write the final
+  // object in end(), so we need to remember the start position.
+  // 2) If this is the field of an object, we use startWritePos to tell our parent the field's
+  // offset within the encoded parent object.
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:103]: I think these things should be handled by the builder. I don't think the converter code needs to deal with variant encoding.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 25]
```diff
@@ -0,0 +1,799 @@
...
+        if (!field.isPrimitive() || field.asPrimitiveType().getPrimitiveTypeName() != BINARY) {
+          throw new IllegalArgumentException("Value must be a binary value");
+        }
+      } else if (fieldName.equals("typed_value")) {
+        this.typedValueIdx = i;
+      }
+    }
+
+    if (valueIdx >= 0) {
+      converters[valueIdx] = new VariantValueConverter(this);
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:154]: If there are unexpected fields in the schema, this could result in an `ArrayIndexOutOfBoundsException`.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 26]
```diff
@@ -0,0 +1,799 @@
...
+    }
+
+    if (typedValueIdx >= 0) {
+      Converter typedConverter = null;
+      Type field = fields.get(typedValueIdx);
+      LogicalTypeAnnotation annotation = field.getLogicalTypeAnnotation();
+      if (annotation instanceof LogicalTypeAnnotation.ListLogicalTypeAnnotation) {
+        typedConverter = new VariantArrayConverter(field.asGroupType());
+      } else if (!field.isPrimitive()) {
+        GroupType typed_value = field.asGroupType();
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:164]: Nit: Normally you'd continue to follow Java conventions and use a camelCase `typedValue` rather than using the column name.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java [hunk 27]
```diff
@@ -0,0 +1,799 @@
...
+    // Record the offset of this element in the binary.
+    parentConverter.addElement();
+  }
+
+  @Override
+  public void end() {}
+}
+
+class VariantObjectConverter extends GroupConverter implements VariantConverter {
+  private VariantBuilderHolder builder;
```
Comment (rdblue) [inline parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:738]: Calling this `builder` results in later calls to `builder.builder.getWritePos()`. It would be better to name this `holder`.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java (hunk 25 above)
Comment (cashmand) [reply parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:154]: Oops, good catch.

File: parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java (hunk 24 above)
Comment (cashmand) [reply parquet-variant/src/main/java/org/apache/parquet/variant/VariantColumnConverter.java:103]: This is mostly fixed when I move to the new builder API. The old interface required offsets to be managed by the caller.

File: parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java (hunk 19 above)
Comment (cashmand) [reply parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java:711]: I think it was from the original version of Gene's PR that I was working on top of. It's gone after I rebased over the latest version.
